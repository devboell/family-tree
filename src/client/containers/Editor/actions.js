/* eslint-disable import/prefer-default-export */
import {
  SET_CREATE_MODE,
} from './actionTypes'


export const setCreateMode = flag => ({
  type: SET_CREATE_MODE,
  flag,
})
