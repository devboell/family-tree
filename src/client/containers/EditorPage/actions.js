import {
  SET_SELECTED_PERSON_ID,
  SET_CREATE_MODE,
  SET_IS_REMOVING,
} from './constants'

export const setSelectedPersonId = id => ({
  type: SET_SELECTED_PERSON_ID,
  id,
})

export const setCreateMode = flag => ({
  type: SET_CREATE_MODE,
  flag,
})

export const setIsRemoving = flag => ({
  type: SET_IS_REMOVING,
  flag,
})

export const clearSelection = () => ({
  type: SET_SELECTED_PERSON_ID,
  id: 'no_selection',
})
