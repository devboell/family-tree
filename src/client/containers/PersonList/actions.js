import {
  SET_SELECTED_PERSON_ID,
  SET_SELECTION_MODE,
} from './actionTypes'

export const setSelectedPersonId = id => ({
  type: SET_SELECTED_PERSON_ID,
  id,
})

export const setSelectionMode = mode => ({
  type: SET_SELECTION_MODE,
  mode,
})
