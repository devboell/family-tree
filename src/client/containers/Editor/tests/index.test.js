import React from 'react'
import selectionModes from 'containers/PersonList/constants'
import {
  Editor,
  convertFormValuesToSchema,
  newPerson,
  mapStateToProps,
  mapDispatchToProps,
} from '../index'
import { setCreateMode } from '../actions'

describe('Editor component', () => {
  let props
  const formValues = {
    id: 1,
    partners: [
      {
        partner: {
          id: 2,
        },
        relationId: 1,
      },
    ],
  }

  const createId = 1

  beforeEach(() => {
    props = {
      // from enhancer
      loading: false,
      relations: [],
      createPersonMutation: jest.fn(() =>
        Promise.resolve({ data: { createPerson: { id: createId } } })),
      updatePersonMutation: jest.fn().mockReturnValue(Promise.resolve()),
      removePersonMutation: jest.fn(),

      // from parent
      persons: [],
      person: {},
      refetch: jest.fn().mockReturnValue(Promise.resolve()),
      selectionMode: 'UNSELECTED',
      onSelectPerson: jest.fn(),
      onSelectFirst: jest.fn(),

      // from redux
      createMode: true,
      onSetCreateMode: jest.fn(),
      onPrepareCreatePerson: jest.fn(),
    }
  })

  it('convertFormValuesToSchema', () => {
    const expected = {
      id: 1,
      partners: [
        {
          relationId: 1,
          partnerId: 2,
        },
      ],
    }

    expect(convertFormValuesToSchema(formValues)).toEqual(expected)
  })

  it('maps state to props', () => {
    const state = {
      editor: {
        createMode: true,
      },
    }

    const stateProps = mapStateToProps(state)
    expect(stateProps.createMode).toBe(true)
  })

  describe('CREATE mode', () => {
    it('it renders without crashing', () => {
      expect(shallow(<Editor {...props} />)).toMatchSnapshot()
    })

    it('calls createPersonSubmit', async () => {
      const wrapper = shallow(<Editor {...props} />)
      const formProps = wrapper.find('ReduxForm').props()
      const onSubmit = formProps.onSubmit

      await onSubmit(formValues)

      expect(onSubmit.name).toBe('createPersonSubmit')
      expect(props.onSelectPerson).toHaveBeenCalledWith(createId)
    })

    it('sets showRemove prop of EditorControls to false', async () => {
      const wrapper = shallow(<Editor {...props} />)
      const controlProps = wrapper.find('EditorControls').props()

      expect(controlProps.showRemove).toBe(false)
    })

    it('passes newPerson as initialValues to PersonEditor', async () => {
      const wrapper = shallow(<Editor {...props} />)
      const formProps = wrapper.find('ReduxForm').props()

      expect(formProps.initialValues).toEqual(newPerson)
    })

    it('sets createMode to false when selectionMode is not UNSELECTED', async () => {
      const wrapper = shallow(<Editor {...props} />)
      const instance = wrapper.instance()
      const nextProps = {
        onSetCreateMode: jest.fn(),
        selectionMode: selectionModes.SELECTED,
      }
      instance.componentWillReceiveProps(nextProps)
      expect(nextProps.onSetCreateMode).toHaveBeenCalledWith(false)
    })

    it('keeps createMode unchanged when selectionMode is UNSELECTED', async () => {
      const wrapper = shallow(<Editor {...props} />)
      const instance = wrapper.instance()
      const nextProps = {
        onSetCreateMode: jest.fn(),
        selectionMode: selectionModes.UNSELECTED,
      }
      instance.componentWillReceiveProps(nextProps)
      expect(nextProps.onSetCreateMode).not.toHaveBeenCalled()
    })
  })

  describe('UPDATE mode', () => {
    beforeEach(() => {
      props = {
        ...props,
        createMode: false,
        person: {
          id: 1,
          partners: [],
        },
        selectionMode: 'SELECTED',
      }
    })
    it('it renders without crashing', () => {
      expect(shallow(<Editor {...props} />)).toMatchSnapshot()
    })

    it('calls updatePersonSubmit', async () => {
      const wrapper = shallow(<Editor {...props} />)
      const formProps = wrapper.find('ReduxForm').props()
      const onSubmit = formProps.onSubmit

      await onSubmit(formValues)

      expect(onSubmit.name).toBe('updatePersonSubmit')
    })

    it('calls removePersonSubmit', async () => {
      const wrapper = shallow(<Editor {...props} />)
      const formProps = wrapper.find('EditorControls').props()
      const onRemovePerson = formProps.onRemovePerson

      await onRemovePerson()

      expect(props.removePersonMutation).toHaveBeenCalledWith({ id: props.person.id })
      expect(props.onSelectFirst).toHaveBeenCalled()
    })

    it('sets showRemove prop of EditorControls to true', async () => {
      const wrapper = shallow(<Editor {...props} />)
      const controlProps = wrapper.find('EditorControls').props()

      expect(controlProps.showRemove).toBe(true)
    })

    it('passes person as initialValues to PersonEditor', async () => {
      const wrapper = shallow(<Editor {...props} />)
      const formProps = wrapper.find('ReduxForm').props()

      expect(formProps.initialValues).toEqual(props.person)
    })
  })

  describe('LOADING mode', () => {
    beforeEach(() => {
      props = {
        ...props,
        loading: true,
      }
    })
    it('renders loading ... div', () => {
      expect(shallow(<Editor {...props} />)).toMatchSnapshot()
    })
  })

  describe('EMPTY mode, showEditor is false', () => {
    beforeEach(() => {
      props = {
        ...props,
        createMode: false,
      }
    })
    it('only renders EditorControls', () => {
      expect(shallow(<Editor {...props} />)).toMatchSnapshot()
    })
  })

  describe('mapDispatchToProps', () => {
    const dispatch = jest.fn()
    const ownProps = {
      onUnselect: jest.fn(),
    }

    const {
      onSetCreateMode,
      onPrepareCreatePerson,
    } = mapDispatchToProps(dispatch, ownProps)

    beforeEach(() => {
      dispatch.mockReset()
    })

    it('maps onSetCreateMode to setCreateMode', () => {
      const flag = true
      const expected = setCreateMode(flag)
      onSetCreateMode(flag)
      expect(dispatch).toHaveBeenCalledWith(expected)
    })

    it('maps onPrepareCreatePerson to onUnselect and setCreateMode', () => {
      const flag = true
      const expected = setCreateMode(flag)
      onPrepareCreatePerson(flag)
      expect(ownProps.onUnselect).toHaveBeenCalled()
      expect(dispatch).toHaveBeenCalledWith(expected)
    })
  })
})
