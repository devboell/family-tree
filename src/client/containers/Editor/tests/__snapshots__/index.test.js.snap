// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor component CREATE mode it renders without crashing 1`] = `
<styled.div>
  <EditorControls
    onPrepareCreatePerson={[Function]}
    onRemovePerson={[Function]}
    showRemove={false}
  />
  <ReduxForm
    initialValues={
      Object {
        "bornToId": null,
        "name": "",
        "partners": Array [],
      }
    }
    onSubmit={[Function]}
    partners={Array []}
    persons={Array []}
    relations={Array []}
  />
</styled.div>
`;

exports[`Editor component EMPTY mode, showEditor is false only renders EditorControls 1`] = `
<styled.div>
  <EditorControls
    onPrepareCreatePerson={[Function]}
    onRemovePerson={[Function]}
    showRemove={false}
  />
</styled.div>
`;

exports[`Editor component LOADING mode renders loading ... div 1`] = `
<div>
  loading ...
</div>
`;

exports[`Editor component UPDATE mode it renders without crashing 1`] = `
<styled.div>
  <EditorControls
    onPrepareCreatePerson={[Function]}
    onRemovePerson={[Function]}
    showRemove={true}
  />
  <ReduxForm
    initialValues={
      Object {
        "id": 1,
        "partners": Array [],
      }
    }
    onSubmit={[Function]}
    partners={Array []}
    persons={Array []}
    relations={Array []}
  />
</styled.div>
`;
